src/DataProtection/DataProtection/src/KeyManagement/KeyRingProvider.cs

```diff
@@ -292,6 +311,146 @@ internal IKeyRing GetCurrentKeyRingCore(DateTime utcNow, bool forceRefresh = fal
         }
     }
 
+    private IKeyRing GetCurrentKeyRingCoreNew(DateTime utcNow, bool forceRefresh)
+    {
+        Debug.Assert(utcNow.Kind == DateTimeKind.Utc);
+
+        // The 99% and perf-critical case is that there is no task in-flight and the cached
+        // key ring is valid.  We do what we can to avoid unnecessary overhead (locking,
+        // context switching, etc) on this path.
+
+        CacheableKeyRing? existingCacheableKeyRing = null;
+
+        // Can we return the cached keyring to the caller?
+        if (!forceRefresh)
+        {
+            existingCacheableKeyRing = Volatile.Read(ref _cacheableKeyRing);
+            if (CacheableKeyRing.IsValid(existingCacheableKeyRing, utcNow))
+            {
+                return existingCacheableKeyRing.KeyRing;
+            }
+        }
+
+        // If work kicked off by a previous caller has completed, we should use those results
+        // We check this outside the lock to reduce contention in the common case (no task).
+        // Logically, it would probably make more sense to check this before checking whether
+        // the cache is valid - there could be a newer value available - but keeping that path
+        // fast is more important.  The next forced refresh or cache expiration will cause the
+        // new value to be picked up.
+        var existingTask = Volatile.Read(ref _cacheableKeyRingTask);
```

Comment:
I could only think of contrived situations where this could be available while the cache is still valid, but I may well be missing something.

Reply:
I think that's an artifact from when this check happened before the cache read above.  I agree that it's pointless now.

Reply:
See https://github.com/dotnet/aspnetcore/pull/55295